### PR TITLE
API: implement __array_function__ for ExtensionArray

### PR DESCRIPTION
**Pull request by simonjayhawkins**
xref #26380

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thanks. I think this should probably be a mixin that arrays can opt into by inheriting from it.

**Comment by simonjayhawkins** [comment]
for out internal EAs we could opt-in for all. It maybe that we don't need to opt-in for PandasArray and just opt-in for StringArray

**Comment by simonjayhawkins** [comment]
> It maybe that we don't need to opt-in for PandasArray and just opt-in for StringArray

opt-in for a subclass of a class that doesn't opt-in seems more trouble than it's worth since would need to override some of the inherited methods. 

so will add tests back for PandasArray and add ArrayFunctionMixin shortly.

File: pandas/core/arrays/numpy_.py [hunk 1]
```diff
@@ -347,13 +348,17 @@ def min(self, skipna: bool = True, **kwargs) -> Scalar:
...
+    amin = min
+
     def max(self, skipna: bool = True, **kwargs) -> Scalar:
         nv.validate_max((), kwargs)
         result = masked_reductions.max(
             values=self.to_numpy(), mask=self.isna(), skipna=skipna
         )
         return result
 
+    amax = max
```
Comment (jbrockmendel) [inline pandas/core/arrays/numpy_.py:360]: could go on NDArrayBackedExtensionArrayMixin?

Comment (simonjayhawkins) [reply pandas/core/arrays/numpy_.py]: so to keep the changes minimal have used `ea_func = getattr(type(self), func.__name__, None)` in `__array_function__` to start.

but being explicit and adding `@implements(np.amin)` to `def min()` may be a better solution. (may also want separate _HANDLED_FUNCTIONS dicts for each array)

Comment (jbrockmendel) [reply pandas/core/arrays/numpy_.py]: makes sense, thanks

**Comment by simonjayhawkins** [comment]
tests failing

```

        msg = "axis 1 is out of bounds for array of dimension 1"
        with pytest.raises(np.AxisError, match=msg):
>           np.concatenate([data] * 3, axis=1)
E           Failed: DID NOT RAISE <class 'numpy.AxisError'>
```

the changes to make this work was broken off in #35038 but reverted in #36115

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -37,6 +37,41 @@
...
+    def decorator(func):
+        _HANDLED_FUNCTIONS[numpy_function] = func
+        return func
+
+    return decorator
+
+
+class ArrayFunctionMixin:
+    def __array_function__(self, func, types, args, kwargs):
+        if func not in _HANDLED_FUNCTIONS:
```
Comment (jbrockmendel) [inline pandas/core/arrays/base.py:57]: is _HANDLED_FUNCTIONS supposed to be global or class-specific?

Comment (jbrockmendel) [reply pandas/core/arrays/base.py:57]: never mind, i think i get it

Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:57]: I think that needs discussion. it is global here in order to get the basic scaffold in place to implement \_\_array_function__ and get ci to pass (ex. https://github.com/pandas-dev/pandas/pull/35032#issuecomment-691866304) . but for consistency with \_HANDLED_TYPES in \_\_array_ufunc__ we may want to make this class-specific to allow decorating array class methods or have class specific compat methods. concat_compat is compatible with all extensions arrays so decorating this just works (ex. https://github.com/pandas-dev/pandas/pull/35032#issuecomment-691866304).

**Comment by simonjayhawkins** [comment]
> Thanks. I think this should probably be a mixin that arrays can opt into by inheriting from it.

after some more thought I not so sure this is a good idea, otherwise we won't be able to use the numpy api in the codebase in the future to avoid special casing EAs in the code which was my motivation for investigating this.

of course, this would be an additional burden on 3rd party EA authors so needs further consideration, xref #32586

**Pull request closed without merging**